planner/cascades/stringer.go

```diff
@@ -0,0 +1,99 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cascades
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+
+	"github.com/pingcap/tidb/planner/memo"
+)
+
+// ToString stringifies a Group Tree.
+func ToString(g *memo.Group) []string {
+	idMap := make(map[*memo.Group]int)
+	idMap[g] = 0
+	return toString(g, idMap, map[*memo.Group]struct{}{}, []string{})
+}
+
+// toString recursively stringifies a Group Tree using a preorder traversal method.
+func toString(g *memo.Group, idMap map[*memo.Group]int, visited map[*memo.Group]struct{}, strs []string) []string {
+	if _, exists := visited[g]; exists {
+		return strs
+	}
+	visited[g] = struct{}{}
+	// Add new Groups to idMap.
+	for item := g.Equivalents.Front(); item != nil; item = item.Next() {
+		expr := item.Value.(*memo.GroupExpr)
+		for _, childGroup := range expr.Children {
+			if _, exists := idMap[childGroup]; !exists {
+				idMap[childGroup] = len(idMap)
+			}
+		}
+	}
+	// Visit self first.
+	strs = append(strs, groupToString(g, idMap)...)
+	// Visit children then.
+	for item := g.Equivalents.Front(); item != nil; item = item.Next() {
+		expr := item.Value.(*memo.GroupExpr)
+		for _, childGroup := range expr.Children {
+			strs = toString(childGroup, idMap, visited, strs)
+		}
+	}
+	return strs
+}
+
+// groupToString only stringifies a single Group.
+// Format:
+// Group#1 Columns:[a, b, c]
+//     Selection_13 input:Group#2 gt(a, b)
+//     Projection_15 input:Group#3 a, b, c
+func groupToString(g *memo.Group, idMap map[*memo.Group]int) []string {
+	result := make([]string, 0, g.Equivalents.Len()+1)
+	result = append(result, fmt.Sprintf("Group#%d %s", idMap[g], g.Prop.Schema.String()))
+	for item := g.Equivalents.Front(); item != nil; item = item.Next() {
+		expr := item.Value.(*memo.GroupExpr)
+		result = append(result, "    "+groupExprToString(expr, idMap))
```

Comment:
Is the `\t` better than four blank spaces?

Reply:
I remember different editor shows `\t` differently. Some editors may show it using 8 blank spaces (it seems can be set in vim). The space useage is not important than the visual effect for Explain, I think.

Reply:
I vote for spaces

Reply:
Fine.